### ISSUE TITLE
sendrawtransaction support

### ISSUE THREAD

**Issue by lllwvlvwlll** [CityOfZion/neon-js#592]
This feature incorporates support for the sendrawtransaction method.  Initial focus will be on native tokens.

**Comment by ixje** [comment]
The `sendrawtransaction` RPC method exists and works as intended. What's arguably missing is adding helper function to ensure that the transaction we're trying to send it constructed in a valid fashion. For example;
* make sure it is signed
* make sure the first Signer in the `signers` property is the sender of the transaction 
* make sure all except the first signer cannot have a scope of `FeeOnly`
* make sure network fee is calculated after after a sender has been added to the tx
* make sure the max transaction size does not exceed 102400
* make sure the `validUntilBlock` property does not exceed current_height + MaxValidUntilBlockIncrement

**Comment by snowypowers** [comment]
so previously in NEO2, we had the 3 big methods: sendAsset, doInvoke and claimGas. These were super high level methods (as in they accept one gigantic config object and it would just assemble the transaction as intended). I think here we are trying to replicate the sendAsset method.

I have been thinking of redesigning it since the flows are a lot more aligned in NEO3. Will write more about it later.

**Comment by ixje** [comment]
To my best knowledge `sendrawtransaction` is still the only RPC call that allows the TX to be persisted to the chain. That makes sending assets just one use case for `sendrawtransaction`. Smart contract function invocations would be another. Lastly, (I've never seen this, but possible) any computation using the blockchain as it just executes the `script` attached to the transaction. It does not necessarily have to call an existing smart contract. 

I tried to reason an interface from a consumer perspective (without prior `neon-js` experience) and my first rough idea was something as follows;

1. api.nep5. (see https://github.com/CityOfZion/neon-js/issues/595)
   - getName()
   - getTotalSupply()
   - getSymbol()
   - getDecimals()
   - getBalanceOf()
   - transfer()

2. api.NEO (extends api.nep5)
   - getUnclaimedGas()
   - claimGas()

3. api.GAS (extends api.nep5)

4. api.Contract.
   - dynamically expose whatever methods the abi exposes

and then use as follows;
```javascript
// random NEP5 contract usage
const my_contract = api.nep5("0xmycontract_script_hash");
my_contract.transfer(from, to, amount);

// NEO contract usage
api.NEO.transfer(from, to, amount)
api.NEO.getUnclaimedGas();
api.NEO.claimGas();

// custom smart contract usage
const my_cool_contract = api.Contract("0xmy_cool_contract_script_hash");
my_cool_contract.cool_function1();
```

**Comment by snowypowers** [comment]
tldr: https://github.com/CityOfZion/neon-js/blob/next/docs/guides/basic/transfer.md is the rough draft of what I think the flow can be.

With everything as a contract, the flow is pretty streamlined into:

1. Assembling the script
2. Assemble Transaction
3. Verifying script contents (sender has enough assets, script works with invokescript, etc) 
4. Sign Transaction
5. sendrawtransaction

In Neo2, the flow was common with the only difference being the different Transaction types to use and different fields in the transaction to fill. For Neo3, it is even easier with only 1 transaction type. We can provide different script generators instead.

So an example:
```
export function claimGas(claimGasOptions,commonConfig) {
const script = generateScript(claimGasOptions);
const tx = generateTx(script, commonConfig);
validateTx(tx);
signTx(tx, commonConfig);
sendRawTransaction(tx, commonConfig);
}
```

**Comment by ixje** [comment]
I got a couple of questions

> tldr: https://github.com/CityOfZion/neon-js/blob/next/docs/guides/basic/transfer.md is the rough draft of what I think the flow can be.

1. I can't tell from the reply if your intention is to have the code of `transfer.md` to be internal code or not? 
From an end user perspective that seem pretty low level and could be simplified a lot for them by a higher level wrappers e.g. as in the interface I described above.  

> With everything as a contract, the flow is pretty streamlined into:
> 
> 1. Assembling the script
> 2. Assemble Transaction
> 3. Verifying script contents (sender has enough assets, script works with invokescript, etc)
> 4. Sign Transaction
> 5. sendrawtransaction
> 
I agree so far on this order

> In Neo2, the flow was common with the only difference being the different Transaction types to use and different fields in the transaction to fill. For Neo3, it is even easier with only 1 transaction type. We can provide different script generators instead.
> 
> So an example:
> 
> ```
> export function claimGas(claimGasOptions,commonConfig) {
> const script = generateScript(claimGasOptions);
> const tx = generateTx(script, commonConfig);
> validateTx(tx);
> signTx(tx, commonConfig);
> sendRawTransaction(tx, commonConfig);
> }
> ```

2. I'm going to take a guess on the `claimGasOptions` and `commonConfig` contents as I don't see those mentioned in the `transfer.md` linked.
   - `commonConfig`
      - RPC address
      - network magic
      - optional extra network fee for transaction prioritisation
   - `claimGasOptions`
      - `from` address 
      - `to` address

   Anything missing?

3. What do you have in mind on how to expose `claimGas()` and alike? Like so `api.NEO.claimGas();` ? or differently?
4. Seeing the `transfer.md` file it seems like (nearly?) all the functionality is there to create a NEO3 compatible release of `neon-js`, right? Or what am I missing? Please let me know what you think at this moment needs to be solved for a NEO3 release.

Finally, one thing I think we should change from the `transfer.md` is getting the current balance via the NEP5Tracker plugin. It will unnecessarily limit the amount of nodes that can be used. The easy alternative is an `invokefunction` with a `balanceOf` call that's available on all RPC nodes.

**Comment by snowypowers** [comment]
1. Yes, the intent is to wrap the contents of `transfer.md` into a higher level abstraction that is available within `neon-js`. This should be available as both a single method call (claimGas()) and the individual steps so that a custom workflow can be crafted.

2. You pretty much cover almost all bases. Maybe also `blocksTillExpiry` as an extra indicator for users to indicate how long they want the transaction to last. These are details that I will leave to the implementor's discretion. I believe the current options you listed are sufficient for a first pass.

3. Likely as standalone methods instead of tying them to specific namespaces. I dont think its too much of a hassle for developers to pass in the contracts as parameters. So `api.claimGas()` and `api.transferToken()` are favorable.

4. Will answer in #599 since that seems like the better place for plans.


Yes on moving away from NEP5Tracker plugin. That was just done cos I wanted to test my integration :P Should be trivial.

**Comment by lllwvlvwlll** [comment]
resolved
